### PR TITLE
cli: Add cockroach dump support for --dump-mode=schema

### LINKED ISSUE #61756: sql/cli: make `cockroach dump --dump-mode=schema` delegate to SHOW CREATE ALL TABLES
We want to make sure the use case described in https://github.com/cockroachdb/cockroach/issues/53488 is not broken.

In 21.2, we should remove the `dump` CLI command entirely.

cc  @RichardJCai that means you don't need to add anything to `master` -- just the `release-21.1` branch.

### PR DESCRIPTION
**Pull request by RichardJCai**
cli: Add cockroach dump support for --dump-mode=schema

Release note (cli change): Add back support for cockroach dump --dump-mode=schema.
This command calls SHOW CREATE ALL TABLES and returns the output.
This gives users a migration path to use in 21.1 while switching over
to SHOW CREATE ALL TABLES for 21.2.
When executing this, the user is warned that the command is being removed
in 21.2 and they should use SHOW CREATE ALL TABLES instead.

This is only intended to be backported to 21.1 and not added to master.

Resolves #61756

### CODE REVIEW

**Comment by RichardJCai** [comment]
Requesting review from @adityamaru since you were the one who created the following pr https://github.com/cockroachdb/cockroach/pull/56964

File: pkg/cli/dump.go
Comment (rafiss) [inline pkg/cli/dump.go:39]: no need for this todo if we don't merge this to master :)

but also no need to remove it from this PR

**Review by rafiss — APPROVED** [review]
looks good from my end! but would be good to have bulk IO look

**Review by adityamaru — CHANGES_REQUESTED** [review]
> Reviewed 1 of 6 files at r1.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/61782)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @knz, @rafiss, and @RichardJCai)

*[a discussion](https://reviewable.io/reviews/cockroachdb/cockroach/61782#-MVWmIGQAG42un6y7MvH:-MVWmIGQAG42un6y7MvI:b-6wpy4m) (no related file):*
I think there are a few more places we can tighten up the only supported mode being --schema.

- `flags_util.go` has consts defined for all modes. We can delete all but schema?
- The above will uncover some additional flag parsing methods that support all 3 modes. I believe fixing this will catch an unsupported mode sooner than in `runDumpSchema`?
- `cliflags/flags.go` has a description that is printed on --help. This should be changed to only mention schema.

---
___
*[pkg/cli/context.go, line 351 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/61782#-MVWp8si8Hi9ZCBedgiZ:-MVWp8si8Hi9ZCBedgi_:bifnh4) ([raw file](https://github.com/cockroachdb/cockroach/blob/20e8bd9c4a6690d7a023d75f93dd9099d19c7a83/pkg/cli/context.go#L351)):*
> ```Go
> // test that exercises command-line parsing.
> func setDumpContextDefaults() {
> 	dumpCtx.dumpMode = dumpBoth
> ```

I think we should default to schema, if no args are provided? This will be the only option once the above comments are addressed.
___
*[pkg/cli/dump.go, line 39 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/61782#-MVS8tiP_i6F1ohvwM8J-r1-39:-MVWkJHtDCSOT3vipMfv:bx7fuc4) ([raw file](https://github.com/cockroachdb/cockroach/blob/78e8fcbf85e38e5019176726c8091bbb16c83fb9/pkg/cli/dump.go#L39)):*
<details><summary><i>Previously, rafiss (Rafi Shamim) wrote…</i></summary><blockquote>

no need for this todo if we don't merge this to master :)

but also no need to remove it from this PR
</blockquote></details>

I'm actually a little confused by this comment, runDumpSchema is the driver behind cockroach dump <database> so it isn't really replacing it? I think we can just do away with this comment :shrug: 
___
*[pkg/cli/dump_test.go, line 21 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/61782#-MVWpXXLDlamBSNnFDMk:-MVWpXXM3kB14m9-1ykP:b-nr0ou) ([raw file](https://github.com/cockroachdb/cockroach/blob/20e8bd9c4a6690d7a023d75f93dd9099d19c7a83/pkg/cli/dump_test.go#L21)):*
> ```Go
> )
> 
> func TestDatabaseDumpCommand(t *testing.T) {
> ```

I assume `SHOW CREATE ALL TABLES` has relevant tests, but would it be worth adding some more cases. Off the top of my head:
- UDT
- UDS
- constraints
- column family
- temp tables

I'll defer to you if you have covered all these cases in `SHOW CREATE ALL TABLES` already :slightly_smiling_face: 
___
*[pkg/cli/dump_test.go, line 75 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/61782#-MVWckiq9K53UTVSpI8s:-MVWckirAPRQSNwUdBFU:bykwrjy) ([raw file](https://github.com/cockroachdb/cockroach/blob/20e8bd9c4a6690d7a023d75f93dd9099d19c7a83/pkg/cli/dump_test.go#L75)):*
> ```
> 			if err != nil {
> 				t.Fatal(err)
> 			}
> ```
nit: here and below, can we use require.NoError(t, err)
___
*[pkg/cli/dump_test.go, line 97 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/61782#-MVWcRDFE8KYatnIECBq:-MVWcRDFE8KYatnIECBr:bc02881) ([raw file](https://github.com/cockroachdb/cockroach/blob/20e8bd9c4a6690d7a023d75f93dd9099d19c7a83/pkg/cli/dump_test.go#L97)):*
> ```
> 			if dump != recreatedDump {
> 				t.Fatalf("expected: %s\ngot: %s", dump, recreatedDump)
> 			}
> ```
can we use require.Equal(...)?
___
*[pkg/cli/dump_test.go, line 99 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/61782#-MVWhKNkCQstBOgPBFaW:-MVWhKNkCQstBOgPBFaX:b-izaj42) ([raw file](https://github.com/cockroachdb/cockroach/blob/20e8bd9c4a6690d7a023d75f93dd9099d19c7a83/pkg/cli/dump_test.go#L99)):*
> ```Go
> 			if dump != recreatedDump {
> 				t.Fatalf("expected: %s\ngot: %s", dump, recreatedDump)
> 			}
> ```

Just to ensure the round trip-ability, can we also `select *` on the tables and compare output. Similar to https://github.com/cockroachdb/cockroach/pull/56964/files#diff-8d89379fd4d1c6d3c583b3bdc89738be1b210566b83a262b145126dca44ce145L618
___
*[pkg/cli/dump_test.go, line 144 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/61782#-MVWhYTe-wE7E5qD7Tvs:-MVWhYTe-wE7E5qD7Tvt:b-rnxd9f) ([raw file](https://github.com/cockroachdb/cockroach/blob/20e8bd9c4a6690d7a023d75f93dd9099d19c7a83/pkg/cli/dump_test.go#L144)):*
> ```Go
> 	for _, test := range tests {
> 		out, err := c.RunWithCaptureArgs(test.cmd)
> 		if err != nil {
> ```

require.NoError(t, err)
___
*[pkg/cli/dump_test.go, line 148 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/61782#-MVWh_si4fm0KTKkcg4B:-MVWcgTE6ePQVeG8wvIl:b-8yfhqo) ([raw file](https://github.com/cockroachdb/cockroach/blob/20e8bd9c4a6690d7a023d75f93dd9099d19c7a83/pkg/cli/dump_test.go#L148)):*
> ```
> 		if out != test.expected {
> 			t.Errorf("test failed: %s\n %s", test.name, pretty.Diff(test.expected, out))
> 		}
>
require.Equal(t, ...)

**Review by RichardJCai — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/61782)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @adityamaru, @knz, and @rafiss)

*[a discussion](https://reviewable.io/reviews/cockroachdb/cockroach/61782#-MVWmIGQAG42un6y7MvH:-MVX20JiDr6e4DGK1Ey8:b7es4r2) (no related file):*
<details><summary><i>Previously, adityamaru (Aditya Maru) wrote…</i></summary><blockquote>

I think there are a few more places we can tighten up the only supported mode being --schema.

- `flags_util.go` has consts defined for all modes. We can delete all but schema?
- The above will uncover some additional flag parsing methods that support all 3 modes. I believe fixing this will catch an unsupported mode sooner than in `runDumpSchema`?
- `cliflags/flags.go` has a description that is printed on --help. This should be changed to only mention schema.
</blockquote></details>

I deleted all but schema, however added back a dumpModeNone as default. The reason being is that previously without specifiying a mode, the user would dump both. I want them to explicitly specify the schema flag so they know they're only dumping schema.

Updated the the help text

---
___
*[pkg/cli/context.go, line 351 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/61782#-MVWp8si8Hi9ZCBedgiZ:-MVX2GJJ8D32jPaAGA--:b-4ib3gq) ([raw file](https://github.com/cockroachdb/cockroach/blob/20e8bd9c4a6690d7a023d75f93dd9099d19c7a83/pkg/cli/context.go#L351)):*
<details><summary><i>Previously, adityamaru (Aditya Maru) wrote…</i></summary><blockquote>

I think we should default to schema, if no args are provided? This will be the only option once the above comments are addressed.
</blockquote></details>

Changed it to default to dumpModeNone which will error, want to make sure the user knows they're dumping schema only.
___
*[pkg/cli/dump.go, line 39 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/61782#-MVS8tiP_i6F1ohvwM8J-r1-39:-MVX2Roq4YoxQP92JL6X:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/78e8fcbf85e38e5019176726c8091bbb16c83fb9/pkg/cli/dump.go#L39)):*
<details><summary><i>Previously, adityamaru (Aditya Maru) wrote…</i></summary><blockquote>

I'm actually a little confused by this comment, runDumpSchema is the driver behind cockroach dump <database> so it isn't really replacing it? I think we can just do away with this comment :shrug: 
</blockquote></details>

Done.
___
*[pkg/cli/dump_test.go, line 21 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/61782#-MVWpXXLDlamBSNnFDMk:-MVX2SJd8c8ZXKxziAN6:b-51674i) ([raw file](https://github.com/cockroachdb/cockroach/blob/20e8bd9c4a6690d7a023d75f93dd9099d19c7a83/pkg/cli/dump_test.go#L21)):*
<details><summary><i>Previously, adityamaru (Aditya Maru) wrote…</i></summary><blockquote>

I assume `SHOW CREATE ALL TABLES` has relevant tests, but would it be worth adding some more cases. Off the top of my head:
- UDT
- UDS
- constraints
- column family
- temp tables

I'll defer to you if you have covered all these cases in `SHOW CREATE ALL TABLES` already :slightly_smiling_face: 
</blockquote></details>

I was missing tests for UDT and column families but I added that it on the second commit which I'll merge to master.
___
*[pkg/cli/dump_test.go, line 97 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/61782#-MVWcRDFE8KYatnIECBq:-MVX2X1yBfbufHYk3dIm:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/20e8bd9c4a6690d7a023d75f93dd9099d19c7a83/pkg/cli/dump_test.go#L97)):*
<details><summary><i>Previously, adityamaru (Aditya Maru) wrote…</i></summary><blockquote>

> ```
> 			if dump != recreatedDump {
> 				t.Fatalf("expected: %s\ngot: %s", dump, recreatedDump)
> 			}
> ```
can we use require.Equal(...)?
</blockquote></details>

Done.
___
*[pkg/cli/dump_test.go, line 99 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/61782#-MVWhKNkCQstBOgPBFaW:-MVX2XYa2P3mdGoGyMor:b-dij11m) ([raw file](https://github.com/cockroachdb/cockroach/blob/20e8bd9c4a6690d7a023d75f93dd9099d19c7a83/pkg/cli/dump_test.go#L99)):*
<details><summary><i>Previously, adityamaru (Aditya Maru) wrote…</i></summary><blockquote>

Just to ensure the round trip-ability, can we also `select *` on the tables and compare output. Similar to https://github.com/cockroachdb/cockroach/pull/56964/files#diff-8d89379fd4d1c6d3c583b3bdc89738be1b210566b83a262b145126dca44ce145L618
</blockquote></details>

select * won't return any rows, not sure if that's useful in this case.
___
*[pkg/cli/dump_test.go, line 144 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/61782#-MVWhYTe-wE7E5qD7Tvs:-MVX2cvZEdTJk4pPAN4M:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/20e8bd9c4a6690d7a023d75f93dd9099d19c7a83/pkg/cli/dump_test.go#L144)):*
<details><summary><i>Previously, adityamaru (Aditya Maru) wrote…</i></summary><blockquote>

require.NoError(t, err)
</blockquote></details>

Done.
___
*[pkg/cli/dump_test.go, line 148 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/61782#-MVWh_si4fm0KTKkcg4B:-MVX2dITE4hsQ35GvnXg:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/20e8bd9c4a6690d7a023d75f93dd9099d19c7a83/pkg/cli/dump_test.go#L148)):*
<details><summary><i>Previously, adityamaru (Aditya Maru) wrote…</i></summary><blockquote>

> ```
> 		if out != test.expected {
> 			t.Errorf("test failed: %s\n %s", test.name, pretty.Diff(test.expected, out))
> 		}
>
require.Equal(t, ...)
</blockquote></details>

Done.

**Review by adityamaru — CHANGES_REQUESTED** [review]
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/61782)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @adityamaru, @knz, @rafiss, and @RichardJCai)

___
*[pkg/cli/context.go, line 351 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/61782#-MVWp8si8Hi9ZCBedgiZ:-MVXL3ezDcGYGLGIDZ7T:b1ubcge) ([raw file](https://github.com/cockroachdb/cockroach/blob/20e8bd9c4a6690d7a023d75f93dd9099d19c7a83/pkg/cli/context.go#L351)):*
<details><summary><i>Previously, RichardJCai (Richard Cai) wrote…</i></summary><blockquote>

Changed it to default to dumpModeNone which will error, want to make sure the user knows they're dumping schema only.
</blockquote></details>

Makes sense
___
*[pkg/cli/dump.go, line 24 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/61782#-MVXJiiXBf48CM9-jy6P:-MVXJiiXBf48CM9-jy6Q:b-cel2eo) ([raw file](https://github.com/cockroachdb/cockroach/blob/b3d8ceee461c69a2bf16e74ac8fe54805b2103d3/pkg/cli/dump.go#L24)):*
> ```Go
> // dumpCmd dumps SQL tables.
> var dumpCmd = &cobra.Command{
> 	Use:   "dump [options] <database> --dump-mode=schema",
> ```

I think what we want is to update the help text in `cliflags/flags.go` to only mention schema, and leave this as it was. If you run `./cockroach dump --help` you currently see:

```
Usage:
  cockroach dump [options] <database> --dump-mode=schema [flags]
```

which is slightly confusing cause --dump-mode is a flag and is listed in the flags description printed below. 
___
*[pkg/cli/dump_test.go, line 21 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/61782#-MVWpXXLDlamBSNnFDMk:-MVXNmRW4UlKQwOhLQTD:b-ucfknh) ([raw file](https://github.com/cockroachdb/cockroach/blob/20e8bd9c4a6690d7a023d75f93dd9099d19c7a83/pkg/cli/dump_test.go#L21)):*
<details><summary><i>Previously, RichardJCai (Richard Cai) wrote…</i></summary><blockquote>

I was missing tests for UDT and column families but I added that it on the second commit which I'll merge to master.
</blockquote></details>

Thanks!
___
*[pkg/cli/dump_test.go, line 99 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/61782#-MVWhKNkCQstBOgPBFaW:-MVXIa6ABG5q7ss4Lart:b-carmty) ([raw file](https://github.com/cockroachdb/cockroach/blob/20e8bd9c4a6690d7a023d75f93dd9099d19c7a83/pkg/cli/dump_test.go#L99)):*
<details><summary><i>Previously, RichardJCai (Richard Cai) wrote…</i></summary><blockquote>

select * won't return any rows, not sure if that's useful in this case.
</blockquote></details>

good point :thumbsup: 
___
*[pkg/cli/flags_util.go, line 137 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/61782#-MVXL92t1oKnn6TeuRsy:-MVXL92t1oKnn6TeuRsz:b-suizt2) ([raw file](https://github.com/cockroachdb/cockroach/blob/b3d8ceee461c69a2bf16e74ac8fe54805b2103d3/pkg/cli/flags_util.go#L137)):*
> ```Go
> 	dumpSchemaOnly dumpMode = iota
> 	//dumpModeNone is set by default and will error if it is set.
> 	dumpModeNone
> ```

nit: the type already has "mode" in it, lets call this dumpNone and make this the iota since it is the default.

**Review by RichardJCai — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/61782)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @adityamaru, @knz, @rafiss, and @RichardJCai)

___
*[pkg/cli/context.go, line 351 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/61782#-MVWp8si8Hi9ZCBedgiZ:-MVXRfJcBPmHNqNCyES6:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/20e8bd9c4a6690d7a023d75f93dd9099d19c7a83/pkg/cli/context.go#L351)):*
<details><summary><i>Previously, adityamaru (Aditya Maru) wrote…</i></summary><blockquote>

Makes sense
</blockquote></details>

Done.
___
*[pkg/cli/dump.go, line 24 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/61782#-MVXJiiXBf48CM9-jy6P:-MVXRfuC8HZcFixKZh6L:b-id5dou) ([raw file](https://github.com/cockroachdb/cockroach/blob/b3d8ceee461c69a2bf16e74ac8fe54805b2103d3/pkg/cli/dump.go#L24)):*
<details><summary><i>Previously, adityamaru (Aditya Maru) wrote…</i></summary><blockquote>

I think what we want is to update the help text in `cliflags/flags.go` to only mention schema, and leave this as it was. If you run `./cockroach dump --help` you currently see:

```
Usage:
  cockroach dump [options] <database> --dump-mode=schema [flags]
```

which is slightly confusing cause --dump-mode is a flag and is listed in the flags description printed below. 
</blockquote></details>

Gotcha, updated

**Review by adityamaru — APPROVED** [review]
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/61782)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @adityamaru, @knz, @rafiss, and @RichardJCai)

___
*[pkg/cli/flags_util.go, line 135 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/61782#-MVXjOtV1s8EoElULS23:-MVXjOtV1s8EoElULS24:b-fo8wgq) ([raw file](https://github.com/cockroachdb/cockroach/blob/6bc361fba225cf3aa3dda898eaf330e43f113e59/pkg/cli/flags_util.go#L135)):*
> ```Go
> 
> const (
> 	//dumpModeNone is set by default and will error if it is set.
> ```

nit: dumpNone 
since you'll have to push anyways after regenerating the bazel BUILD files.
___
*[pkg/cli/cliflags/flags.go, line 229 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/61782#-MVXjcNK5jp1AYRci45q:-MVXjcNK5jp1AYRci45r:bflkxk2) ([raw file](https://github.com/cockroachdb/cockroach/blob/6bc361fba225cf3aa3dda898eaf330e43f113e59/pkg/cli/cliflags/flags.go#L229)):*
> ```Go
> 		Name: "dump-mode",
> 		Description: `
> What to dump. Only --dump-mode=schema is supported.`,
> ```

nit: "schema" dumps the schema only. It is the only supported dump-mode.

**Comment by adityamaru** [comment]
Just to confirm:
Commit 1 goes into 21.1.
Commit 2 goes into both master and 21.1.

**Comment by RichardJCai** [comment]
Backport here, will open new PR to merge the logictest.

**Comment by RichardJCai** [comment]
Closing this, refer to 
https://github.com/cockroachdb/cockroach/pull/61871
and 
https://github.com/cockroachdb/cockroach/pull/61870

**Pull request closed without merging**
